expression/builtin_time.go

```diff
@@ -896,51 +897,71 @@ func (c *weekFunctionClass) getFunction(args []Expression, ctx context.Context)
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinWeekSig{newBaseBuiltinFunc(args, ctx)}
+
+	argTps := []evalTp{tpTime}
+	if len(args) == 2 {
+		argTps = append(argTps, tpInt)
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, argTps...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	bf.tp.Flen, bf.tp.Decimal = 2, 0
+
+	var sig builtinFunc
+	if len(args) == 2 {
+		sig = &builtinWeekWithModeSig{baseIntBuiltinFunc{bf}}
+	} else {
+		sig = &builtinWeekWithoutModeSig{baseIntBuiltinFunc{bf}}
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinWeekSig struct {
-	baseBuiltinFunc
+type builtinWeekWithModeSig struct {
+	baseIntBuiltinFunc
 }
 
-func (b *builtinWeekSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+// evalInt evals WEEK(date, mode).
+// see: https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_week
+func (b *builtinWeekWithModeSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	date, isNull, err := b.args[0].EvalTime(row, sc)
 	if err != nil {
-		return d, errors.Trace(err)
+		sc.AppendWarning(err)
```

Comment:
We should consider whether it is in a read or write statement. We could leave a TODO here.
mysql> SELECT WEEK('2008-xx',0);
+-------------------+
| WEEK('2008-xx',0) |
+-------------------+
|              NULL |
+-------------------+
1 row in set, 1 warning (0.00 sec)

mysql> insert into t values(WEEK('2008-xx',0));
ERROR 1292 (22007): Incorrect datetime value: '2008-xx'
mysql>